### PR TITLE
add calculate age func; neo4j types

### PR DESCRIPTION
**Pull request by jcskywalker**

### CODE REVIEW

File: pkg/funcs/functions.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+	"github.com/araddon/dateparse"
+	oc "github.com/cloudprivacylabs/opencypher"
+	"github.com/neo4j/neo4j-go-driver/neo4j"
+)
+
+func init() {
+	oc.RegisterGlobalFunc(oc.Function{
+		Name:      "age",
+		MinArgs:   0,
```
Comment (bserdar) [inline pkg/funcs/functions.go]: MinArgs:1

File: pkg/funcs/functions.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+	return oc.RValue{Value: age}, nil
+}
+
+// valueAsNeo4jDate returns the neo4j.Date value. If value is not neo4j.Date,
+// rReturns ErrStringValueRequired. If an err argument is given, that error is returned.
+func valueAsNeo4jDate(v oc.Value, err ...error) (neo4j.Date, error) {
+	if len(err) != 0 {
+		return neo4j.Date{}, err[0]
+	}
+	fmt.Println(reflect.TypeOf(v.Get()))
```
Comment (bserdar) [inline pkg/funcs/functions.go]: Remove

File: pkg/funcs/functions.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+	s, ok := v.Get().(neo4j.Date)
+	if !ok {
+		return neo4j.Date{}, oc.ErrStringValueRequired
```
Comment (bserdar) [inline pkg/funcs/functions.go]: Make a ErrDateValueRequired type

File: pkg/funcs/functions.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+// valueAsNeo4jDateTime returns the neo4j.LocalDateTime value. If value is not neo4j.LocalDateTime,
+// rReturns ErrStringValueRequired. If an err argument is given, that error is returned.
+func valueAsNeo4jDateTime(v oc.Value, err ...error) (neo4j.LocalDateTime, error) {
+	if len(err) != 0 {
+		return neo4j.LocalDateTime{}, err[0]
+	}
+	s, ok := v.Get().(neo4j.LocalDateTime)
+	if !ok {
+		return neo4j.LocalDateTime{}, oc.ErrStringValueRequired
```
Comment (bserdar) [inline pkg/funcs/functions.go]: Make a ErrDateTimeValueRequired type

File: pkg/funcs/functions.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+			tDate, err = dateparse.ParseAny(n4jDate.String())
+			if err != nil {
+				return nil, err
+			}
+		}
+
+	}
+	// age will only represent the year
+	deltaMonths := (tDob.Year()*12 + int(tDob.Month()) - 1) - (tDate.Year()*12 + int(tDate.Month()) - 1)
+	age := int(math.Floor(float64(deltaMonths / 12)))
```
Comment (bserdar) [inline pkg/funcs/functions.go]: Floor(float(x/y)) = x/y, no need for type conversions

File: pkg/funcs/functions.go [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+func ageFunc(ctx *oc.EvalContext, args []oc.Value) (oc.Value, error) {
+	if args[0].Get() == nil {
+		return oc.RValue{}, nil
+	}
+	var n4jDate neo4j.Date
+	var n4jDateTime neo4j.LocalDateTime
+	var err error
+	// check if argument a neo4j.Date or neo4j.LocalDateTime
+	if n4jDate, err = valueAsNeo4jDate(args[0]); err != nil {
```
Comment (bserdar) [inline pkg/funcs/functions.go]: Can't do else here. The first error will terminate. If the input is neo4jdatetime, it will never check it.

File: pkg/funcs/functions.go [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+		return nil, err
+	} else if n4jDateTime, err = valueAsNeo4jDateTime(args[0]); err != nil {
+		return nil, err
+	}
+
+	var tDob time.Time
+	if n4jDateTime != (neo4j.LocalDateTime{}) {
+		tDob, err = dateparse.ParseAny(n4jDateTime.String())
```
Comment (bserdar) [inline pkg/funcs/functions.go]: Don't use dateParse, there is a method in neo4j times to get Go time

File: pkg/funcs/functions.go [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+		}
+	} else {
+		tDob, err = dateparse.ParseAny(n4jDate.String())
+		if err != nil {
+			return nil, err
+		}
+	}
+	tDate := time.Now()
+	if len(args) > 1 && args[1].Get() != nil {
+		if n4jDate, err = valueAsNeo4jDate(args[1]); err != nil {
```
Comment (bserdar) [inline pkg/funcs/functions.go]: Same here, first error will terminate

**Review by bserdar — CHANGES_REQUESTED** [review]

File: go.mod [hunk 9]
```diff
@@ -1,40 +1,76 @@
 module github.com/cloudprivacylabs/leap
 
-go 1.18
+go 1.19
+
+replace github.com/cloudprivacylabs/opencypher => /Users/jasoncruz/opencypher
```
Comment (bserdar) [inline go.mod]: Remove replace

File: pkg/input/dir_test.go [hunk 10]
```diff
@@ -16,11 +16,25 @@ package input
...
 	"io/ioutil"
 	"testing"
+
+	pkg "github.com/cloudprivacylabs/leap/pkg/funcs"
+	"github.com/cloudprivacylabs/lpg"
+	"github.com/cloudprivacylabs/opencypher"
 )
 
 func TestReadDir(t *testing.T) {
+	opencypher.RegisterGlobalFunc(opencypher.Function{
```
Comment (bserdar) [inline pkg/input/dir_test.go]: Why is this test here?

**Review by bserdar — CHANGES_REQUESTED** [review]

File: pkg/funcs/functions.go [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+func AgeFunc(ctx *oc.EvalContext, args []oc.Value) (oc.Value, error) {
+	if args[0].Get() == nil {
+		return oc.RValue{}, nil
+	}
+	var n4jDate neo4j.Date
+	var n4jDateTime neo4j.LocalDateTime
+	var err error
+	// check if argument a neo4j.Date or neo4j.LocalDateTime
```
Comment (bserdar) [inline pkg/funcs/functions.go:33]: Create a time.Time{} at the top, with each if-statement, parse it using one of the neo4j types

goDate:=time.Time{}
if dt, ok:=args[0].Get().(neo4j.DateTime); ok {
    goDate=dt.Time()
}
...


now:=time.Now()
// Same if blocks with the second arg

**Review by bserdar — CHANGES_REQUESTED** [review]

File: pkg/funcs/functions.go [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+	oc.RegisterGlobalFunc(oc.Function{
+		Name:      "age",
+		MinArgs:   1,
+		MaxArgs:   2,
+		ValueFunc: ageFunc,
+	})
+	oc.RegisterGlobalFunc(oc.Function{
+		Name:      "lookupValueset",
+		MinArgs:   1,
+		MaxArgs:   -1,
```
Comment (bserdar) [inline pkg/funcs/functions.go]: MaxArgs:1

File: pkg/funcs/functions.go [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+		return oc.RValue{}, nil
+	}
+	vs, ok := args[0].Get().(map[string]oc.Value)
+	if !ok {
+		return oc.RValue{}, nil
+	}
+	reqs := make([]ls.ValuesetLookupRequest, 0)
+	for k, v := range vs {
+		reqs = append(reqs, ls.ValuesetLookupRequest{
+			TableIDs:  []string{k},
```
Comment (bserdar) [inline pkg/funcs/functions.go]: TableIDs comes from []string{vs["tableId"]}
Rest of the keys in that map will go to KeyValues
You cannot type assert v.Get() to map[string]string.
input; {tableId:"table", x:y, w:z}
request:
TableIDS: []string{"table},
KeyValues: {x:y,w:z}

File: pkg/funcs/functions_test.go [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+	ctx := opencypher.NewEvalContext(lpg.NewGraph())
+	for _, tt := range ageTests {
+		v, err := opencypher.ParseAndEvaluate(tt.query, ctx)
+		if err != nil {
+			t.Error(err)
+		}
+		if rGetter(v) != tt.expected {
+			t.Errorf("Got age: %d, expected age: %d", rGetter(v), tt.expected)
+		}
+	}
```
Comment (bserdar) [inline pkg/funcs/functions_test.go:84]: Add a test for valueSet

**Review by bserdar — CHANGES_REQUESTED** [review]

File: pkg/funcs/functions_test.go [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+		return v.Get().(opencypher.ResultSet).Rows[0]["1"].Get()
+	}
+	ageTests := []ageTest{
+		{
+			query:    `return age(parseDate("01/01/2000", "MM/DD/YYYY"))`,
+			expected: 22,
+		},
+		{
+			query:    `return age(parseDate("01/01/1900", "MM/DD/YYYY"), parseDate("01/01/2022", "MM/DD/YYYY"))`,
+			expected: 122,
```
Comment (bserdar) [inline pkg/funcs/functions_test.go:48]: Add a test case with "03/03/2001" -> "03/03/2010" and "03/03/2001" -> "03/02/2010". That's the edge case.

File: pkg/funcs/functions_test.go [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+	expected int
+}
+
+func TestAge(t *testing.T) {
+	rGetter := func(v opencypher.Value) interface{} {
+		return v.Get().(opencypher.ResultSet).Rows[0]["1"].Get()
+	}
+	ageTests := []ageTest{
+		{
+			query:    `return age(parseDate("01/01/2000", "MM/DD/YYYY"))`,
```
Comment (bserdar) [inline pkg/funcs/functions_test.go]: This test will not age well. It will start failing in January. Add a GetNow() func variable to the functions.go,initialize it to time.Now, change it for the test.

**Review by bserdar — CHANGES_REQUESTED** [review]

File: pkg/funcs/functions.go [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+		ValueFunc: ageFunc,
+	})
+	oc.RegisterGlobalFunc(oc.Function{
+		Name:      "lookupValueset",
+		MinArgs:   1,
+		MaxArgs:   1,
+		ValueFunc: lookupValuesetFunc,
+	})
+}
+
```
Comment (bserdar) [inline pkg/funcs/functions.go:25]: var GetNow=time.Now

File: pkg/funcs/functions.go [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+		return oc.RValue{}, nil
+	}
+	goDate := time.Time{}
+	// check if argument a neo4j.Date or neo4j.LocalDateTime
+	if date, ok := args[0].Get().(neo4j.Date); ok {
+		goDate = date.Time()
+	} else if dateTime, ok := args[0].Get().(neo4j.LocalDateTime); ok {
+		goDate = dateTime.Time()
+	}
+	tm := time.Now()
```
Comment (bserdar) [inline pkg/funcs/functions.go]: tm:=GetNow()

File: pkg/funcs/functions_test.go [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+
+type ageTest struct {
+	query    string
+	expected int
+}
+
+func TestAge(t *testing.T) {
+	rGetter := func(v opencypher.Value) interface{} {
+		return v.Get().(opencypher.ResultSet).Rows[0]["1"].Get()
+	}
```
Comment (bserdar) [inline pkg/funcs/functions_test.go:22]: oldNow:=GetNow()
defer func() { GetNow=oldNow}()
GetNow=func() time.Time { some fixed date}

**Review by bserdar — CHANGES_REQUESTED** [review]

File: pkg/funcs/functions.go [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+	})
+	oc.RegisterGlobalFunc(oc.Function{
+		Name:      "lookupValueset",
+		MinArgs:   1,
+		MaxArgs:   1,
+		ValueFunc: lookupValuesetFunc,
+	})
+}
+
+var GetNow = time.Now()
```
Comment (bserdar) [inline pkg/funcs/functions.go]: No, this should be a function: var GetNow=time.Now

**Review by bserdar — CHANGES_REQUESTED** [review]
